TribalWars/src/Infrastructure/Repository/BuildingRequiredRepository.cs

```diff
@@ -0,0 +1,80 @@
+﻿using ApplicationCore.Entities;
+using ApplicationCore.Enums;
+using ApplicationCore.Interfaces;
+using Infrastructure.Data;
+using Microsoft.EntityFrameworkCore;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Infrastructure.Repository
+{
+    public class BuildingRequiredRepository : BaseRepository, IBuildingRequiredRepository
+    {
+        public BuildingRequiredRepository(ApplicationDbContext context) : base(context)
+        {
+        }
+
+        public bool HasMaterial(int buildingId, int level, int villageId)
+        {
+            var requiredMaterial = RequiredMaterials(level, buildingId);
+            var playerMaterials = _context.VillageMaterials.Where(material => material.VillageId == villageId).ToList();
+            if (playerMaterials.Count < requiredMaterial.Count())
+            {
+                return false;
+            }
+            foreach (var material in playerMaterials)
+            {
+                var required = requiredMaterial.FirstOrDefault(m => m.MaterialId == material.MaterialId);
+                if(material.Quantity < required.Quantity)
+                {
+                    return false;
+                }
+            }
```

Comment:
I think you should check here every material in `foreach(varmaterial in requiredMaterials)`, then, if `playerMaterials` doesn't contain even one of the required ones, you can return false. So you would need to check if foundPlayerMaterial is not null, and then compare quantities

Reply:
`            var requiredMaterial = RequiredMaterials(level, buildingId);
            var playerMaterials = _context.VillageMaterials.Where(material => material.VillageId == villageId).ToList();
            foreach (var required in requiredMaterial)
            {
                var material = playerMaterials.FirstOrDefault(m => m.MaterialId == required.MaterialId);
                if(material == null || material.Quantity < required.Quantity)
                {
                    return false;
                }
            }
            return true;`

You mean somethink like that?